pkg/quota/image/imagestreamtag_evaluator.go

```diff
@@ -0,0 +1,110 @@
+	kapi "k8s.io/kubernetes/pkg/api"
+	kerrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/resource"
+	kquota "k8s.io/kubernetes/pkg/quota"
+	"k8s.io/kubernetes/pkg/runtime"
+
+	osclient "github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	quotautil "github.com/openshift/origin/pkg/quota/util"
+)
+
+const imageStreamTagEvaluatorName = "Evaluator.ImageStreamTag"
+
+// NewImageStreamTagEvaluator computes resource usage of ImageStreamsTags. It's sole purpose is to handle
+// UPDATE admission operations on imageStreamTags resource.
+func NewImageStreamTagEvaluator(osClient osclient.Interface) kquota.Evaluator {
+	computeResources := []kapi.ResourceName{
+		imageapi.ResourceImages,
+	}
+
+	matchesScopeFunc := func(kapi.ResourceQuotaScope, runtime.Object) bool { return true }
+	getFuncByNamespace := func(namespace, id string) (runtime.Object, error) {
+		nameParts := strings.SplitN(id, ":", 2)
+		if len(nameParts) != 2 {
+			return nil, fmt.Errorf("%q is an invalid id for an imagestreamtag. Must be in form <name>:<tag>.", id)
+		}
+
+		obj, err := osClient.ImageStreamTags(namespace).Get(nameParts[0], nameParts[1])
+		if err != nil {
+			if !kerrors.IsNotFound(err) {
+				return nil, err
+			}
+			obj = &imageapi.ImageStreamTag{
+				ObjectMeta: kapi.ObjectMeta{
+					Namespace: namespace,
+					Name:      id,
+				},
+			}
+		}
+		return obj, nil
+	}
+
+	return quotautil.NewSharedContextEvaluator(
+		imageStreamTagEvaluatorName,
+		kapi.Kind("ImageStreamTag"),
+		map[admission.Operation][]kapi.ResourceName{admission.Update: computeResources},
+		computeResources,
+		matchesScopeFunc,
+		getFuncByNamespace,
+		nil,
+		imageStreamConstraintsFunc,
+		makeImageStreamTagUsageComputerFactory(osClient))
+}
+
+// imageStreamTagConstraintsFunc checks that given object is an image stream tag
+func imageStreamTagConstraintsFunc(required []kapi.ResourceName, object runtime.Object) error {
+	if _, ok := object.(*imageapi.ImageStreamTag); !ok {
+		return fmt.Errorf("Unexpected input object %v", object)
+	}
+	return nil
+}
+
+// makeImageStreamTagUsageComputerFactory returns an object used during computation of image quota across all
+// repositories in a namespace.
+func makeImageStreamTagUsageComputerFactory(osClient osclient.Interface) quotautil.UsageComputerFactory {
+	return func() quotautil.UsageComputer {
+		return &imageStreamTagUsageComputer{
+			GenericImageStreamUsageComputer: *NewGenericImageStreamUsageComputer(osClient, false, true),
+		}
+	}
+}
+
+// imageStreamUsageComputer is a context object for use in SharedContextEvaluator.
+type imageStreamTagUsageComputer struct {
+	GenericImageStreamUsageComputer
+}
+
+// Usage returns a usage for an image stream tag.
+func (c *imageStreamTagUsageComputer) Usage(object runtime.Object) kapi.ResourceList {
+	ist, ok := object.(*imageapi.ImageStreamTag)
+	if !ok {
+		return kapi.ResourceList{}
+	}
+
+	img := &ist.Image
```

Comment:
My point is that image streams are not generic blobs, but inputs that
are calculated after admission.  We should check the quota of the
image stream beforehand and then reject if you are over (which it
sounds like you agree with) but we can do that without attempting to
calculate the incoming tag change.

Tagging an image from another namespace or importing it is not a
serious problem for us if the image exceeds the quota (in both cases
size is accounted for).  Only the number of images per stream needs to
be enforced, and only the size of pushed images.  We don't even need
to be exact on the number of images per stream - if we are at 21 and
quota is 20, we're still fine.  Any create can be rejected, and any
delete can be allowed.

That is simpler than trying to do this incremental calculation, and
remove the need to duplicate business logic that only the image stream
should be aware of.

Basically, on create IST, check count of images and reject if over
limit.  Always allow IST delete.  Have replenish only look at IS.